cloudinit/distros/alpine.py

```diff
@@ -202,6 +204,423 @@ def preferred_ntp_clients(self):
+                    # Redact certain fields from the logs
+                    if key in redact_opts:
+                        log_add_cmd.extend([add_opts[key], "REDACTED"])
+                    else:
+                        log_add_cmd.extend([add_opts[key], val])
+
+            elif key in add_flags and val:
+                add_cmd.append(add_flags[key])
+                if use_useradd:
+                    log_add_cmd.append(add_flags[key])
+
+        # Don't create the home directory if directed so or if the user is a
+        # system user
+        if kwargs.get("no_create_home") or kwargs.get("system"):
+            if use_useradd:
+                # Shadow useradd
+                add_cmd.append("-M")
+                log_add_cmd.append("-M")
+            else:
+                # Busybox adduser
+                add_cmd.append("-H")
+        else:
+            if use_useradd:
+                # Shadow useradd
+                add_cmd.append("-m")
+                log_add_cmd.append("-m")
+
+        if not use_useradd:
+            # Busybox 'adduser' puts username at end of command
+            add_cmd.append(name)
+
+        # Run the command
+        LOG.debug("Adding user %s", name)
+        if use_useradd:
+            try:
+                subp.subp(add_cmd, logstring=log_add_cmd)
+            except Exception as e:
+                util.logexc(LOG, "Failed to create user %s", name)
+                raise e
+        else:
+            try:
+                subp.subp(add_cmd)
+            except Exception as e:
+                util.logexc(LOG, "Failed to create user %s", name)
+                raise e
+
+        if not use_useradd:
+            if addn_groups is not None:
+                # Separately add user to each additional group as Busybox
+                #  'adduser' does not support specifying additional groups.
+                for addn_group in addn_groups:
+                    LOG.debug("Adding user to group %s", addn_group)
+                    try:
+                        subp.subp(["addgroup", name, addn_group])
+                    except Exception as e:
+                        util.logexc(
+                            LOG,
+                            "Failed to add user %s to group %s",
+                            name,
+                            addn_group,
+                        )
+                        raise e
+
+            if passwd is not None:
+                # Separately set password as Busybox 'adduser' does not
+                # support passing password as CLI option.
+                self.set_passwd(name, passwd, hashed=True)
+
+            if expirydate is not None or inactive is not None:
+                # Separately set expirydate and/or inactive values directly
+                # in /etc/shadow file as Busybox 'adduser' does not support
+                # passing these as CLI options.
+                shadow_contents = None
+                with open("/etc/shadow", "r") as file:
```

Comment:
Do we want to use `util.load_text_file` here instead of `open` to ensure we log what cloud-init is reading? I't probably not critical as we are logging our write to /etc/shadow below which is indication enough in logs that cloud-init is meddling in /etc/shadow manipulation.

Reply:
I had already (locally) rewritten the code to use util.load_text_file.

I've just pushed my latest changes.